### DISCUSSION TITLE
Using PowerApps CLI in an Azure Runbook on Hybrid Worker

### DISCUSSION THREAD

**Discussion by zhead13** [microsoft/powerplatform-build-tools#612] [Q&A]
I'm trying to use the PowerApps CLI in an Azure Runbook running a Windows Hybrid Worker. The "pac" commands are recognised when directly on the Windows Server, but I cannot get them to be recognized in the Hybrid Runbook.  

When I run **Find-Package -Name "Microsoft.PowerApps.CLI"** it finds it ok, but still fails on any pac commands:
pac : The term 'pac' is not recognized as the name of a cmdlet, function, script file, or operable program.

Do I need to import the package into Runbook session somehow, like you do with Modules?
Any help appreciated.

**Comment by devkeydet** [comment]
It's probably not on the path.  Have you tried `(Get-Command pac).Source`?  If running this command doesn't result in a location, then it's not on the path.  You could add it `$env:PATH += ";PathToPac"` 

An alternative might be to install pac using dotnet.  When you do that, it automatically adds it to the path.
`dotnet tool install --global Microsoft.PowerApps.CLI.Tool`
https://learn.microsoft.com/en-us/power-platform/developer/cli/introduction#install-power-platform-cli-for-windows

**Reply by zhead13** [reply]
Thanks for you reply..No luck,,,Tried both ways installed via dotnet and also tried manually converting the nupkg file to zip and added to $ENV:PSMODULEPATH
Has anyone got this working in an Azure Runbook (Running on Hybrid)?

**Comment by petrochuk** [comment]
Microsoft.PowerApps.CLI nuget package is used purely as a zip file and compiled with full .net framework. If you need to use cross platform version compiled with .Net 6 then use Microsoft.PowerApps.CLI.**Tool**

**Comment by imenos-wouter-forrier** [comment]
@zhead13, @petrochuk,
Did you find a solution? When I run "dotnet tool install --global Microsoft.PowerApps.CLI.Tool" in a runbook, it returns:
```
Since you just installed the .NET SDK, you will need to reopen the Command Prompt window before running the tool you installed.
You can invoke the tool using the following command: pac
Tool 'microsoft.powerapps.cli.tool' (version '1.29.11') was successfully installed.
```
When I run "dotnet tool list --global" or "dotnet tool list" is doesn't show any installed tools...
Is there a way to "restart" the automation account of runbook in Azure Automation after an installation?

**Reply by zhead13** [reply]
I never got it working

**Reply by Manju5137** [reply]
Hello zhead13,

I even tried to use it in the runbook but the modules are missing out in the azure to make it use in the runbook.

**Comment by imenos-wouter-forrier** [comment]
@petrochuk , where can I get that ZIP file?

**Comment by imenos-wouter-forrier** [comment]
@zhead13 , @Manju5137 , did you in the meantime found a solution to run Power Platform CLI in the cloud?
Currently we use Power Automate Desktop to run it unattended, but it's not the ideal solution.

**Comment by Henry-Sir** [comment]
Hello everyone,

as I currently have the same problem, I came across this discussion.
As already written, the `dotnet tool install --global Microsoft.PowerApps.CLI.Tool` does not work.
Installing the MSI does not work either.
(https://learn.microsoft.com/en-us/power-platform/developer/cli/introduction?tabs=windows#install-microsoft-power-platform-cli)

So the only solution is to download the package and use it that way. For this I have built a function that downloads the Nuget package, unpacks it and inserts the path into the environment variable.

After that the `pac` command can be used normally.

```

Function InstallPac{
    # dotnet tool install --global Microsoft.PowerApps.CLI.Tool
    # dotnet tool install --global Microsoft.PowerApps.CLI
    
    Write-Output "Installing Microsoft Power Platform CLI..."
    $pacDownloadUrl = "https://www.nuget.org/api/v2/package/Microsoft.PowerApps.CLI"
    # Expand-Archive need .zip extension https://github.com/actions/runner/issues/1883
    $pacInstallFile = "$($env:TEMP)\$($pacDownloadUrl.Split("/")[-1])" + $(If(("$($env:TEMP)\$($pacDownloadUrl.Split("/")[-1])" -replace '^.*(?=.{4}$)') -ne ".zip"){".zip"})
    $pacInstallFolder = $env:LOCALAPPDATA + "\Microsoft\PowerAppsCLI\"
    $pacToolsFolder = $pacInstallFolder + "tools\"

    Write-Output "Downloading pac Nuget package as zip file..."
    Invoke-WebRequest $pacDownloadUrl -OutFile $pacInstallFile

    Write-Output "Expand pac zip file to $pacInstallFolder"
    Expand-Archive -Path $pacInstallFile -DestinationPath $pacInstallFolder

    Write-Output "Set Enviroment variable PATH"
    # https://community.dynamics.com/blogs/post/?postid=1ccd34d1-70fb-4c16-8c04-8a9aa23a842c
    $env:PATH = $env:PATH + $(If($env:PATH[-1] -ne ";"){";"}) + $pacToolsFolder + ";"
}

```

I hope it helps. 


EDIT:
My runbook runs in Azure (no hybrid worker). I think for the Hybrid Worker it is enough to install pac once and then update the environment variable.

**Reply by imenos-wouter-forrier** [reply]
@Henry-Sir , thanks for sharing, really appreciated.
I have tried to test it but still get an error, see screenshot.
Can you share how you created the runbook and which PS version?
Did you create it in the Azure Portal or via VS Code?

![image](https://github.com/user-attachments/assets/55f4a624-d282-48e7-983c-3596a2bb931c)
![image](https://github.com/user-attachments/assets/85694f13-f220-48fd-b645-3ef6bfa16b4e)

**Reply by imenos-wouter-forrier** [reply]
It's ok, I got it working as follows. Many thanks!
![image](https://github.com/user-attachments/assets/1d4e0890-5251-4fd1-a62d-f386379e2dd3)
